library/std/src/sys/windows/thread_local_dtor.rs

```diff
@@ -1,4 +1,39 @@
+//! Implements thread-local destructors that are not associated with any
+//! particular data.
+
 #![unstable(feature = "thread_local_internals", issue = "none")]
 #![cfg(target_thread_local)]
+use super::c;
+
+// Using a per-thread list avoids the problems in synchronizing global state.
+#[thread_local]
+static mut DESTRUCTORS: Vec<(*mut u8, unsafe extern "C" fn(*mut u8))> = Vec::new();
+
+pub unsafe fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
+    DESTRUCTORS.push((t, dtor));
+}
+
+// See windows/thread_local_keys.rs for an explanation of this callback function.
+// The short version is that all the function pointers in the `.CRT$XL*` array
+// will be called whenever a thread or process starts or ends.
+
+#[link_section = ".CRT$XLD"]
```

Comment:
This all seems pretty reasonable to me! Could the previously-registered "at thread exit" function get a new hook to call into this module though? I think it 'd probably be best to avoid these `#[link_section]` annotations if we can and have them consolidated into one place if possible.

Reply:
Ok, I've used a normal Rust function for running the new dtors and simply made the exit function call it. Is this what you had in mind?